oshdb-api/src/main/java/org/heigit/ohsome/oshdb/api/mapreducer/MapReducer.java

```diff
@@ -754,17 +754,44 @@ protected MapReducer(MapReducer<?> obj) {
     // apply geometry filter as first map function
     final List<MapFunction> remainingMappers = List.copyOf(ret.mappers);
     ret.mappers.clear();
-    if (ret.forClass.equals(OSMContribution.class)) {
-      ret = ret.filter(x -> {
-        OSMContribution c = (OSMContribution) x;
-        return f.applyOSMGeometry(c.getEntityBefore(), c::getGeometryBefore)
-          || f.applyOSMGeometry(c.getEntityAfter(), c::getGeometryAfter);
-      });
-    } else if (ret.forClass.equals(OSMEntitySnapshot.class)) {
-      ret = ret.filter(x -> {
-        OSMEntitySnapshot s = (OSMEntitySnapshot) x;
-        return f.applyOSMGeometry(s.getEntity(), s::getGeometry);
-      });
+    if (this.grouping == Grouping.NONE) {
+      // no grouping -> directly filter using the geometries of the snapshot / contribution
+      if (ret.forClass.equals(OSMEntitySnapshot.class)) {
+        ret = ret.filter(x -> {
+          OSMEntitySnapshot s = (OSMEntitySnapshot) x;
+          return f.applyOSMGeometry(s.getEntity(), s::getGeometry);
+        });
+      } else if (ret.forClass.equals(OSMContribution.class)) {
+        ret = ret.filter(x -> {
+          OSMContribution c = (OSMContribution) x;
+          return f.applyOSMGeometry(c.getEntityBefore(), c::getGeometryBefore)
+              || f.applyOSMGeometry(c.getEntityAfter(), c::getGeometryAfter);
+        });
+      }
+    } else if (this.grouping == Grouping.BY_ID) {
+      // grouping by entity -> filter each list entry individually
+      if (ret.forClass.equals(OSMEntitySnapshot.class)) {
+        @SuppressWarnings("unchecked") MapReducer<X> filteredListMapper = (MapReducer<X>)
+            ret.map(x -> (Collection<OSMEntitySnapshot>) x)
+                .map(snapshots -> snapshots.stream().filter(s ->
+                    s != null && f.applyOSMGeometry(s.getEntity(), s::getGeometry))
+                    .collect(Collectors.toCollection(ArrayList::new)))
+                .filter(snapshots -> !snapshots.isEmpty());
+        ret = filteredListMapper;
+      } else if (ret.forClass.equals(OSMContribution.class)) {
+        @SuppressWarnings("unchecked") MapReducer<X> filteredListMapper = (MapReducer<X>)
+            ret.map(x -> (Collection<OSMContribution>) x)
+                .map(contributions -> contributions.stream()
+                    .filter(c -> c != null
+                        && (f.applyOSMGeometry(c.getEntityBefore(), c::getGeometryBefore)
+                        || f.applyOSMGeometry(c.getEntityAfter(), c::getGeometryAfter)))
+                    .collect(Collectors.toCollection(ArrayList::new)))
+                .filter(contributions -> !contributions.isEmpty());
+        ret = filteredListMapper;
+      }
+    } else {
+      throw new UnsupportedOperationException(
```

Comment:
Is this an Exception that can actually happen? I think no, so I suppose you cannot test it. But we should keep it, if someone extends the `Grouping` enum without adding a case here.

Reply:
no, it can't happen at the moment because the enum only has two possibilities. At first, I had implemented this with a `switch` (with `default` to be the same as `case NONE`), but then sonarlint complained that one shouldn't use switch statements with only two branches :shrug: I thought by adding the "hypothetical" throw statement here, it's clear that something must be added if at some point there would be an additional _grouping_ possibility.

Reply:
So let's keep that in mind for future versions of the oshdb :+1: